### DISCUSSION TITLE
Display Adapter components

### DISCUSSION THREAD

**Discussion by k-ljung** [RudolphRiedel/FT800-FT813#37] [General]
As there is no "Discussion" feature for the Display Adapter repository, I ask the question here.

I was thinking of building a couple of display adapter cards to use with Teensy and later the ESP32 platform, but due to the lack of components it is completely impossible to find the components. Do you know any alternative components that would work and that you can get hold of?

//Kalle

**Comment by RudolphRiedel** [comment]
That "chip crisis" is annoying me.
I know that switching regulators are problematic to find.
I even bought a couple of RT8259GE on Amazon and got, well, at least working regulators but not really RT8259GE.
But the footprint for these regulators are quite common, using an ADP2301AUJZ for example works as well.
And I just ordered a couple XRP7659ISTR-F from Mouser, apart from the lower maximum voltage these look promising.

Anything else missing that I am not yet aware of?

And which Teensy?
I had an issue with the pins of the Teensy 4.1, it would at first only work with the logic-analyzer attached.
In the end I replaced the 10R series resistors for X_MOSI, X_SCK and X_CS with 240R.

**Reply by k-ljung** [reply]
I use Riverdi screens and feed backlight with 5V and plan to use MIC5504-3.3YM5 (U6). Unfortunately I can not find 74LVC2G17 and 74VHC1GT125, nor the CT-1205H-SMT speaker. I also have a hard time to find the TS4148 in 0805 package.

I'm using Tensy 4 for now and I think I found some german post about the issue with the logic-analyzer attached (or not attached).

**Comment by RudolphRiedel** [comment]
Which Riverdi screens do you use?
The EVE4 screens already have a speaker driver on board and it really should work better than the no-effort-stupid-driver I implemented in order to get the occasional "beep" from the display.
And I believe the drawings are updated with the part number for the connector.

Also, which region are you from?
I am from Germany but I get most of my parts from DigiKey or Mouser, some from TME.

Neither the 74LVC2G17 or the 74VHC1GT125 are necessary per se in this case since their main feature is that they have 5V tolerant inputs while beeing supplied with 3.3V.
Well, and these are rather fast.
So you could use different buffers, I only use these everywhere as I am using too many different parts already. :-)

That said, both are available from Mouser:
https://www.mouser.de/ProductDetail/Nexperia/74LVC2G17GV125?qs=me8TqzrmIYV0s3Bq3xwtOg%3D%3D
https://www.mouser.de/ProductDetail/onsemi/MC74VHC1GT125DBVT1G?qs=vLWxofP3U2wy698N2mKkvQ%3D%3D

And for my latest board I switched to a different inductor:
https://www.mouser.de/ProductDetail/Taiyo-Yuden/NR6028T4R7M?qs=PzICbMaShUfL31t2FerArw%3D%3D

The TS4148 is just a general purpose diode and almost any type would do.
But I checked and Mouser has these as well:
https://www.mouser.de/ProductDetail/Taiwan-Semiconductor/TS4148-RYG?qs=C%2F0tY%2F%2FqZe%2Fr0HVt0T5xsg%3D%3D

I am normally not using these with the switching regulators, these merely boost the efficiency when you are going from 5V to 3.3V.

**Comment by k-ljung** [comment]
I am from Sweden and can definitely order from Mouser. I've checked Digi-Key, Farnell, a few Swedish retailers and I'm also pretty sure I've checked Mouser before, but I might get confused.

At the moment I have three different Riverdi displays, two EVE3 and one EVE4 and it is EVE4 that I will use in the future.

All displays are up and running both with your library and the one from Riverdi. My problem is that I have an unstable SPI bus as soon as the SPI cable gets longer. I need to be able to have a SPI cable of 15-20 cm and think I understand that your display adapter would solve it, or do you have any other tips?

**Comment by RudolphRiedel** [comment]
I originally started using glue logic as I was using 5V controllers.
When I did my first design for the EVE displays that included a 3.3V controller, I went with only the 10R series resistors and the 22pF capacitors.
And that did not work, at least not for useable clock speeds since I had to learn that the drive strength at 3.3V was too weak for double-digit SPI clocks over the FFC.
This was really short on the PCB, the controller was close to the FFC connector but still could not drive the lines.
I even was able to do an experiment with two outputs for SCK and MOSI in parallel since the SPI unit on this controller can be routed to a couple of different pin pairs and the logic allows to use switch on SPI on all of them at the same time.
And this worked a whole lot better, driving the cable at 12+MHz just required more mA than one pin could deliver.

CS and PD are fine without logic.
But for SCK, MOSI and MISO I always use extra logic now and apart from the glitch with the Teensy 4 it just works.

This is the design I am using for the RVT101H and RVT70H on the latest board but also with a different board and a EVE3-43G for example:
![grafik](https://user-images.githubusercontent.com/31180093/148661657-90d0cad8-35fc-4260-b421-48f659bc403e.png)

Edit: the 9V for backlight was for the RVT101H and I reduced that to 7,2V in the meantime, for the RVT70H I am using 5V.


I keep this rather small though, this will be the next iteration:
![grafik](https://user-images.githubusercontent.com/31180093/148661707-9e64f40b-fc7c-4143-b03c-e7a922d13a83.png)

So yes, at least to my experience adding buffers really helps with driving the SPI over the given FFC since these can drive the lines a lot better than a normal controller output buffer at 3.3V.
But at the same time these need to be pretty fast since adding additional delays means that MISO is out of sync with SCK for lower frequencies than it would be otherwise.

**Reply by dominicdumont** [reply]
Umm that's interesting and worrying. I have made a board using a PIC32MX driving an EVE-3 (Matrix Orbital EVE3-50G) through a 75 mm FFC + another 55 mm on the board. No resistor and no capacitor. SPI bus is running @ 20MHz and have not seen any issue _yet_.  

I also have an ESP32 prototype connected to an EVE-3 through 75 mm of wirewrap and SPI bus running at 16 MHz and not see any issue either. 

After reading this thread wondering if I should add them... 

What was the behavior of the display that led you to add the buffers? Was it just not working or was there intermittent display issue?

**Comment by RudolphRiedel** [comment]
I just checked the PIC32MX and it can drive more current at 3.0V than the ATSAMC21 I was using could at 3.3V.
And the difference is less than a factor of two, both only drive single digit mA at 3.xV.
So it really depends on the controller family beeing used.
I have a couple of reports of using ESP32 boards without glue logic.
Would I be building an ESP32 board, I would still add the buffers though.

Well, and my adapters are more for use with everything I can find.

>What was the behavior of the display that led you to add the buffers?

It was not working untill I lowered the frequency below 8MHz, maybe 4MHz.
And I tried a couple of things including leaving out the capacitors and using different resistors like 0R0.

**Reply by dominicdumont** [reply]
Thanks for the reply. I will certainly consider adding them at least to the ESP32 board.

**Comment by k-ljung** [comment]
After a couple of months break with the project, I have run a couple of tests to try to understand what the problem is. As I wrote earlier, I experience an unstable SPI connection if I connect the display with an IDC connector with 10cm cable to a Riverdi break out board, from there I have a 15cm FFC cable to the display.
![image](https://user-images.githubusercontent.com/9248053/148696830-a32ba035-edb1-4aa4-b637-ddaba8bae143.png)

I have three different displays and three different PCB cards to test this on, for all tests I run with EVE_Test_Arduino_PlatformIO code.

The displays I test with are:

- RVT43ALBFWC00 (EVE3)
- RVT50UQBNWC00 (EVE3)
- RVT70HSBNWC00-B (EVE4)

### Test 1
The first test I run on Arduino UNO with Riverdi Shield for Arduino, this works well regardless of whether I connect the displays directly with the FCC cable on the shield or via the IDC connector and break out board.
![image](https://user-images.githubusercontent.com/9248053/148696894-82d4f609-4141-4624-93e0-0cc74cf43b00.png)

If I set the backlight to 100% on the 7'' display, the display stops working, but this is expected as the Arduino UNO is not able to power this with power only from USB.

What I think is that Riverdi Shield for Arduino has an NVT2008BQ bidirectional voltage-level translator that might help get a stable signal?

### Test 2
Test two runs on a Teensy 4 on a custom PCB card with ZIF connector and 15cm FCC cable, here I get a relatively stable connection regardless of displays but I have to work with the SPI bus speed. In this test, the power supply to the display comes from the USB connection to the Teensy.
![image](https://user-images.githubusercontent.com/9248053/148696927-efb7fa2f-2403-4114-b00d-53916c5c576d.png)

The 4.3'' display works with SPI speed 1-12MHz and 30MHz, but not 13, 14 and 24-29MHz.
The 5'' display display works with SPI speed 1-8MHz and 30MHz.
The 7'' display works with all SPI speed from 1-40MHz.

### Test 3
Test three runs on a Teensy 4 on a custom PCB card with IDC connector, flat cable to break out board and FCC cable to the display. On this test I get an unstable connection and the result differs between the displays. In this test, the power supply to the display comes from a DC/DC converter.
![image](https://user-images.githubusercontent.com/9248053/148696971-da40c5e5-4385-4235-89a7-01a9d2acfef5.png)

The 7'' display works best and I can use SPI speed from 1-35MHz.
The 5'' works worst and only with SPI speed 10MHz.
The 4.5'' display works with quite a few different SPI speeds.

In order to get any kind of stable connection for the 5'' and 4.3'' displays at all, I need to set:
```
EVE_memWrite16(REG_GPIOX_DIR, 0xffff);
EVE_memWrite16(REG_GPIOX, 0xffff);
```
I do not understand exactly what these two settings do but it has with drive current of output pins.

Since I get such different results and have a hard time evaluating whether it is my PCB layout, cable types or Teensy, I plan to create a Teensy shield with one IDC connector where I can connect to the Riverdi break out board.

I also plan to create a display adapter with IDC connector, buffers and ZIF connection, inspired by the @RudolphRiedel  example above.

**Comment by k-ljung** [comment]
@RudolphRiedel, what do you think about this design for Teensy 4?
![image](https://user-images.githubusercontent.com/9248053/149009044-5d2a0aa7-1acc-4a6e-a12e-cbd31ae65224.png)

**Comment by RudolphRiedel** [comment]
The voltage regulator is missing a capacitor or two and R9 is not necessary, if at all I would increase R3 but it should be fine.
Otherwise it looks good.

**Reply by k-ljung** [reply]
Okay, so it's only R8 I need to get to work for Teensy?

I added two capacitor for the voltage regulator.
![image](https://user-images.githubusercontent.com/9248053/149032363-7a439ae1-904d-47b6-a55b-a75b0b35f153.png)

**Comment by RudolphRrr** [comment]
R7 and R8 are what I needed at 240R to make a Teensy 4.0 work with my adapter.
And also a 22pF to GND after it to form a low-pass filter.
You have to try this, maybe you can use a different value or just a bridge for the resistors.

And the output capacitor at your LDO might not be correct.
I do not know which part from which manufacturer this is exactly but I briefly checked TI and ON and for both the cap needs to be at least 10µF and since this is over temperature I recommend using 22µF.
And the capacitor needs to be within an ESR range of 0.3 to 22R or 0.33 to 2.2R depending on the exact model of LDO.

**Comment by k-ljung** [comment]
I added C7 and C8, was this what you meant?
I also checked the data sheet for LM1117 and it should be 10uP tantal capacitator.
![image](https://user-images.githubusercontent.com/9248053/149141203-a34b1e04-156a-44e6-b510-e3f003c480c6.png)

**Comment by RudolphRrr** [comment]
>I added C7 and C8, was this what you meant?
Yes, exactly, it may work without these and 0R0 for R7 and R8 but leaving out parts is easier than to add parts afterwards. :-)

>I also checked the data sheet for LM1117 and it should be 10uP tantal capacitator.

Have you checked the ESR as well?
And as I wrote, a 10µF does not always actually have 10µF. :-)
What make and model is the LM1117-3.3?

**Reply by k-ljung** [reply]
I'm thinking of using this one: https://www.electrokit.com/uploads/productfile/40351/lm1117.pdf
This is the data sheet for tantal https://www.electrokit.com/uploads/productfile/40526/T494-9.pdf

**Reply by k-ljung** [reply]
From my understanding, this capacitor have ESR 1.0, I'm using 20V version.

**Comment by RudolphRrr** [comment]
It should work either way but this is hopefully not the datasheet for the LDO you will receive as National Semiconductor was sold to TI in 2011. :-)
And the cap is looking good, 10µF/20V are four parts from 0.5R to 3.0R ESR.
